server/grpc_service.go

```diff
@@ -1969,28 +1969,46 @@ func (s *GrpcServer) WatchGlobalConfig(req *pdpb.WatchGlobalConfigRequest, serve
 	// If the revision is compacted, will meet required revision has been compacted error.
 	// - If required revision < CompactRevision, we need to reload all configs to avoid losing data.
 	// - If required revision >= CompactRevision, just keep watching.
-	watchChan := s.client.Watch(ctx, configPath, clientv3.WithPrefix(), clientv3.WithRev(revision))
+	// Use WithPrevKV() to get the previous key-value pair when get Delete Event.
+	watchChan := s.client.Watch(ctx, configPath, clientv3.WithPrefix(), clientv3.WithRev(revision), clientv3.WithPrevKV())
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
 		case res := <-watchChan:
-			if revision < res.CompactRevision {
-				if err := server.Send(&pdpb.WatchGlobalConfigResponse{
-					Revision: res.CompactRevision,
-					Header: s.wrapErrorToHeader(pdpb.ErrorType_DATA_COMPACTED,
-						fmt.Sprintf("required watch revision: %d is smaller than current compact/min revision. %d", revision, res.CompactRevision)),
-				}); err != nil {
+			if res.Err() != nil {
+				var resp pdpb.WatchGlobalConfigResponse
+				if revision < res.CompactRevision {
+					resp.Header = s.wrapErrorToHeader(pdpb.ErrorType_DATA_COMPACTED,
+						fmt.Sprintf("required watch revision: %d is smaller than current compact/min revision %d.", revision, res.CompactRevision))
+				} else {
+					resp.Header = s.wrapErrorToHeader(pdpb.ErrorType_UNKNOWN,
+						fmt.Sprintf("watch channel meet other error %s.", res.Err().Error()))
+				}
+				if err := server.Send(&resp); err != nil {
 					return err
 				}
+				// Err() indicates that this WatchResponse holds a channel-closing error.
+				return res.Err()
```

Comment:
We should remove this return. return `res.Err` will let grpc deliver this error to the client and close the client, but we already notify the client about this error in the last sent message, so we'd better let the client to choose what to do next, in some situation, the client can just go on without retry creating a new client.

Reply:
for these context-cancel errors there is no need to receive the channel, do you mean to return nil here?

Reply:
ok. That make sense, I think return res.Err is ok here.